File: files/kelly_criterion.py
"""
Kelly Criterion & Capital Management
======================================
Implements:
  - Full Kelly Criterion
  - Fractional Kelly (Half, Quarter, Eighth)
  - Simultaneous bet allocation with covariance adjustment
  - Maximum drawdown protection

Reference:
  - Kelly, J.L. (1956) — "A New Interpretation of Information Rate"
"""

from __future__ import annotations
import numpy as np
from dataclasses import dataclass
from typing import Optional


@dataclass
class BetProposal:
    """A single bet under evaluation."""
    name: str
    true_probability: float     # Model's true probability (0-1)
    decimal_odds: float         # Bookmaker decimal odds
    correlation_with_active: float = 0.0   # Pearson correlation vs active bets

    @property
    def net_odds(self) -> float:
        """b: net profit per unit staked (decimal - 1)."""
        return self.decimal_odds - 1.0

    @property
    def implied_probability(self) -> float:
        return 1.0 / self.decimal_odds

    @property
    def edge(self) -> float:
        """Expected value edge = true_prob * net_odds - (1 - true_prob)."""
        return self.true_probability * self.net_odds - (1.0 - self.true_probability)

    @property
    def expected_value(self) -> float:
        """EV = (true_prob × decimal_odds) - 1."""
        return self.true_probability * self.decimal_odds - 1.0


@dataclass
class KellyOutput:
    """Result of Kelly calculation for one bet."""
    bet_name: str
    full_kelly_fraction: float
    half_kelly_fraction: float
    quarter_kelly_fraction: float
    eighth_kelly_fraction: float
    edge_pct: float
    ev: float
    is_positive_ev: bool
    recommended_fraction: str   # which fraction to use
    recommended_value: float


def kelly_fraction(true_prob: float, net_odds: float) -> float:
    """
    Core Kelly formula: f* = (b*p - q) / b
    where b = net odds, p = true prob, q = 1 - p
    """
    q = 1.0 - true_prob
    if net_odds <= 0:
        return 0.0
    f = (net_odds * true_prob - q) / net_odds
    return max(0.0, f)   # Never bet negative Kelly


def fractional_kelly(bet: BetProposal, fraction: float = 0.25) -> float:
    """Apply fractional Kelly to control variance."""
    f_full = kelly_fraction(bet.true_probability, bet.net_odds)
    return f_full * fraction


class KellyManager:
    """
    Capital management engine using Fractional Kelly criterion.
    """

    def __init__(
        self,
        bankroll: float,
        default_fraction: float = 0.25,   # Quarter-Kelly default
        ev_threshold: float = 0.05,
        max_single_bet_pct: float = 0.05, # Never more than 5% on one bet
    ):
        self.bankroll         = bankroll
        self.default_fraction = default_fraction
        self.ev_threshold     = ev_threshold
        self.max_single_pct   = max_single_bet_pct

    def evaluate(self, bet: BetProposal) -> KellyOutput:
        """Evaluate a single bet proposal."""
        f_full  = kelly_fraction(bet.true_probability, bet.net_odds)
        f_half  = f_full * 0.50
        f_qtr   = f_full * 0.25
        f_eighth= f_full * 0.125

        # Reduce further if highly correlated with active portfolio
        corr_discount = max(0.5, 1.0 - abs(bet.correlation_with_active))
        recommended   = min(f_qtr * corr_discount, self.max_single_pct)

        return KellyOutput(
            bet_name              = bet.name,
            full_kelly_fraction   = round(f_full, 6),
            half_kelly_fraction   = round(f_half, 6),
            quarter_kelly_fraction= round(f_qtr, 6),
            eighth_kelly_fraction = round(f_eighth, 6),
            edge_pct              = round(bet.edge * 100, 4),
            ev                    = round(bet.expected_value, 4),
            is_positive_ev        = bet.expected_value > self.ev_threshold,
            recommended_fraction  = f"Quarter-Kelly (corr-adjusted)",
            recommended_value     = round(recommended, 6),
        )

    def stake_in_dollars(self, output: KellyOutput) -> float:
        """Convert Kelly fraction to dollar stake."""
        return round(self.bankroll * output.recommended_value, 2)

    def evaluate_portfolio(self, bets: list[BetProposal]) -> list[KellyOutput]:
        """
        Evaluate a portfolio of bets with covariance adjustment.
        Only executes bets above EV threshold.
        """
        results = []
        for bet in bets:
            out = self.evaluate(bet)
            if out.is_positive_ev:
                results.append(out)
                print(
                    f"  [+EV] {bet.name:<35} "
                    f"Edge: {out.edge_pct:+.2f}% | "
                    f"Stake: {self.stake_in_dollars(out):.2f} "
                    f"({out.recommended_value*100:.3f}% bankroll)"
                )
            else:
                print(f"  [---] {bet.name:<35} EV={out.ev:+.4f} — SKIP")
        return results


if __name__ == "__main__":
    manager = KellyManager(bankroll=10_000, default_fraction=0.25, ev_threshold=0.05)

    bets = [
        BetProposal("Brazil Win",        true_probability=0.72, decimal_odds=1.65),
        BetProposal("Over 2.5 Goals",    true_probability=0.58, decimal_odds=1.85),
        BetProposal("Scotland Win",      true_probability=0.08, decimal_odds=5.50),
    ]

    print("\n" + "─"*75)
    print("  KELLY CAPITAL MANAGEMENT — Portfolio Evaluation")
    print("─"*75)
    results = manager.evaluate_portfolio(bets)

File: lib/kelly-criterion.js
/**
 * BigFish — Kelly Criterion Calculator
 * 
 * Implements the Fractional Kelly Criterion for optimal bankroll management.
 * The Kelly formula maximizes long-term log-wealth growth while protecting
 * against ruin via fractional staking strategies.
 * 
 * Formula: f* = (b×p - q) / b
 * where b = net decimal odds, p = true win probability, q = 1-p
 * 
 * References:
 * - Kelly (1956) "A New Interpretation of Information Rate"
 * - Thorp (2006) "The Kelly Criterion in Blackjack, Sports Betting..."
 */

import { KELLY_DEFAULTS } from './constants.js';

/**
 * Calculate the full Kelly stake fraction
 * 
 * @param {number} trueProbability - True probability of winning (0-1)
 * @param {number} decimalOdds - Decimal odds offered by bookmaker
 * @returns {number} Optimal fraction of bankroll to wager (can be negative = don't bet)
 */
export function fullKelly(trueProbability, decimalOdds) {
  const b = decimalOdds - 1; // Net decimal odds (profit per unit)
  const p = trueProbability;
  const q = 1 - p;
  
  if (b <= 0) return 0;
  
  const fraction = (b * p - q) / b;
  return fraction;
}

/**
 * Calculate Fractional Kelly stake
 * 
 * @param {number} trueProbability - True win probability
 * @param {number} decimalOdds - Decimal odds
 * @param {number} fraction - Kelly fraction (0.25 = Quarter-Kelly, 0.5 = Half-Kelly)
 * @param {number} bankroll - Current bankroll amount
 * @param {number} maxBetFraction - Maximum bet as fraction of bankroll
 * @returns {Object} Stake recommendation
 */
export function fractionalKelly(
  trueProbability,
  decimalOdds,
  fraction = KELLY_DEFAULTS.FRACTION,
  bankroll = KELLY_DEFAULTS.INITIAL_BANKROLL,
  maxBetFraction = KELLY_DEFAULTS.MAX_BET_FRACTION
) {
  const fullK = fullKelly(trueProbability, decimalOdds);
  const fractionalK = fullK * fraction;
  
  // Don't bet if edge is negative
  if (fractionalK <= 0) {
    return {
      shouldBet: false,
      stake: 0,
      stakeAmount: 0,
      kellyFraction: fullK,
      fractionalKelly: fractionalK,
      edge: calculateEdge(trueProbability, decimalOdds),
      expectedValue: calculateEV(trueProbability, decimalOdds),
      reason: 'Negative edge — no bet recommended',
    };
  }
  
  // Cap at maximum bet fraction
  const cappedFraction = Math.min(fractionalK, maxBetFraction);
  const stakeAmount = cappedFraction * bankroll;
  
  const edge = calculateEdge(trueProbability, decimalOdds);
  const ev = calculateEV(trueProbability, decimalOdds);
  
  return {
    shouldBet: ev > KELLY_DEFAULTS.MIN_EV_THRESHOLD,
    stake: cappedFraction,
    stakeAmount: Math.round(stakeAmount * 100) / 100,
    kellyFraction: fullK,
    fractionalKelly: fractionalK,
    kellyMode: fraction === 1 ? 'Full Kelly' : 
               fraction === 0.5 ? 'Half Kelly' : 
               fraction === 0.25 ? 'Quarter Kelly' : 
               fraction === 0.125 ? 'Eighth Kelly' : 
               `${fraction * 100}% Kelly`,
    edge,
    expectedValue: ev,
    potentialProfit: Math.round(stakeAmount * (decimalOdds - 1) * 100) / 100,
    potentialLoss: stakeAmount,
    growthRate: calculateGrowthRate(trueProbability, decimalOdds, cappedFraction),
    reason: ev > KELLY_DEFAULTS.MIN_EV_THRESHOLD 
      ? `+EV opportunity detected (${(ev * 100).toFixed(1)}% edge)` 
      : `Edge below minimum threshold (${(ev * 100).toFixed(1)}% < ${(KELLY_DEFAULTS.MIN_EV_THRESHOLD * 100)}%)`,
  };
}

/**
 * Calculate Expected Value
 * EV = (probability × decimal_odds) - 1
 * 
 * @param {number} trueProbability
 * @param {number} decimalOdds
 * @returns {number} Expected value per unit staked
 */
export function calculateEV(trueProbability, decimalOdds) {
  return trueProbability * decimalOdds - 1;
}

/**
 * Calculate edge (true probability vs implied)
 * @param {number} trueProbability
 * @param {number} decimalOdds
 * @returns {number} Edge as decimal
 */
export function calculateEdge(trueProbability, decimalOdds) {
  const impliedProb = 1 / decimalOdds;
  return trueProbability - impliedProb;
}

/**
 * Calculate expected log-growth rate (Kelly growth rate)
 * G(f) = p × ln(1 + f×b) + q × ln(1 - f)
 * 
 * @param {number} p - True probability
 * @param {number} odds - Decimal odds
 * @param {number} f - Fraction of bankroll wagered
 * @returns {number} Expected growth rate
 */
export function calculateGrowthRate(p, odds, f) {
  const b = odds - 1;
  const q = 1 - p;
  if (f <= 0 || f >= 1) return 0;
  return p * Math.log(1 + f * b) + q * Math.log(1 - f);
}

/**
 * Calculate risk of ruin given a series of bets
 * Approximation: RoR ≈ ((1-edge)/edge)^(bankroll/avgBetSize)
 * 
 * @param {number} edge - Average edge per bet
 * @param {number} bankroll - Current bankroll
 * @param {number} avgBetSize - Average bet size
 * @returns {number} Approximate probability of ruin (0-1)
 */
export function riskOfRuin(edge, bankroll, avgBetSize) {
  if (edge <= 0) return 1;
  const ratio = (1 - edge) / edge;
  const units = bankroll / avgBetSize;
  return Math.pow(ratio, units);
}

/**
 * Calculate optimal stake for a parlay (multiple correlated legs)
 * Adjusts Kelly for the correlation between legs
 * 
 * @param {number} jointProbability - True joint probability of parlay hitting
 * @param {number} parlayOdds - Combined decimal odds of the parlay
 * @param {number} fraction - Kelly fraction to use
 * @param {number} bankroll - Current bankroll
 * @returns {Object} Parlay stake recommendation
 */
export function parlayKelly(jointProbability, parlayOdds, fraction = KELLY_DEFAULTS.FRACTION, bankroll = KELLY_DEFAULTS.INITIAL_BANKROLL) {
  return fractionalKelly(jointProbability, parlayOdds, fraction, bankroll);
}

/**
 * Simultaneous bet allocation — adjusts stakes when multiple bets are active
 * Accounts for covariance across bets to prevent overexposure
 * 
 * @param {Object[]} bets - Array of bet objects with { trueProbability, decimalOdds }
 * @param {number} bankroll - Current bankroll
 * @param {number} fraction - Kelly fraction
 * @param {number} maxTotal - Maximum total exposure as fraction of bankroll
 * @returns {Object[]} Adjusted stake recommendations
 */
export function simultaneousAllocation(
  bets,
  bankroll = KELLY_DEFAULTS.INITIAL_BANKROLL,
  fraction = KELLY_DEFAULTS.FRACTION,
  maxTotal = 0.15 // Max 15% of bankroll exposed simultaneously
) {
  // Calculate individual Kelly stakes
  const stakes = bets.map(bet => 
    fractionalKelly(bet.trueProbability, bet.decimalOdds, fraction, bankroll)
  );
  
  // Calculate total exposure
  const totalExposure = stakes.reduce((sum, s) => sum + s.stake, 0);
  
  // If total exceeds max, scale down proportionally
  if (totalExposure > maxTotal) {
    const scaleFactor = maxTotal / totalExposure;
    return stakes.map(s => ({
      ...s,
      stake: s.stake * scaleFactor,
      stakeAmount: Math.round(s.stakeAmount * scaleFactor * 100) / 100,
      scaled: true,
      scaleFactor,
    }));
  }
  
  return stakes.map(s => ({ ...s, scaled: false, scaleFactor: 1 }));
}

/**
 * Track bankroll history and calculate performance metrics
 * 
 * @param {Object[]} betHistory - Array of completed bets
 * @param {number} initialBankroll - Starting bankroll
 * @returns {Object} Performance analytics
 */
export function calculatePerformance(betHistory, initialBankroll = KELLY_DEFAULTS.INITIAL_BANKROLL) {
  let bankroll = initialBankroll;
  const history = [{ time: 0, bankroll: initialBankroll }];
  let wins = 0, losses = 0;
  let totalStaked = 0, totalReturn = 0;
  let maxBankroll = initialBankroll, maxDrawdown = 0;
  
  betHistory.forEach((bet, i) => {
    totalStaked += bet.stakeAmount;
    if (bet.won) {
      wins++;
      const profit = bet.stakeAmount * (bet.decimalOdds - 1);
      bankroll += profit;
      totalReturn += bet.stakeAmount + profit;
    } else {
      losses++;
      bankroll -= bet.stakeAmount;
      totalReturn += 0;
    }
    
    maxBankroll = Math.max(maxBankroll, bankroll);
    const drawdown = (maxBankroll - bankroll) / maxBankroll;
    maxDrawdown = Math.max(maxDrawdown, drawdown);
    
    history.push({ time: i + 1, bankroll: Math.round(bankroll * 100) / 100 });
  });
  
  return {
    currentBankroll: Math.round(bankroll * 100) / 100,
    totalBets: betHistory.length,
    wins,
    losses,
    winRate: betHistory.length > 0 ? wins / betHistory.length : 0,
    roi: totalStaked > 0 ? (totalReturn - totalStaked) / totalStaked : 0,
    profitLoss: Math.round((bankroll - initialBankroll) * 100) / 100,
    maxDrawdown,
    sharpeRatio: calculateSharpe(betHistory),
    history,
  };
}

/**
 * Calculate Sharpe ratio of betting performance
 */
function calculateSharpe(betHistory) {
  if (betHistory.length < 2) return 0;
  
  const returns = betHistory.map(bet => {
    if (bet.won) return bet.decimalOdds - 1;
    return -1;
  });
  
  const meanReturn = returns.reduce((s, r) => s + r, 0) / returns.length;
  const variance = returns.reduce((s, r) => s + (r - meanReturn) ** 2, 0) / returns.length;
  const stdDev = Math.sqrt(variance);
  
  return stdDev > 0 ? meanReturn / stdDev : 0;
}
